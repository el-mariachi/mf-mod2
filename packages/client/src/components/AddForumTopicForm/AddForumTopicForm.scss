.add-forum-topic-form {
  textarea {
    height: 200px;
  }
}
