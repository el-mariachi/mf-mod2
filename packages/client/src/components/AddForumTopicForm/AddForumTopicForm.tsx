import Button from 'react-bootstrap/Button'
import Form from 'react-bootstrap/Form'
function AddForumTopicForm() {
  return (
    <Form className="add-topic-form form d-flex align-items-center gap-2">
      <Form.Group controlId="formNewTopic">
        <Form.Control type="text" placeholder="Название новой темы" />
      </Form.Group>
      <Button variant="primary" type="submit">
        Создать тему
      </Button>
    </Form>
  )
}

export default AddForumTopicForm
