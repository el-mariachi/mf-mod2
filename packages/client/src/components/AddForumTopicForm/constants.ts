import { minMax } from '@utils/validations'

const inputData = [
  {
    name: 'topic_name',
    label: 'Название',
    type: 'text',
    placeholder: 'Назовите тему',
    validate: { minMax: minMax(3, 64) },
  },
  {
    name: 'topic_text',
    label: 'Содержимое',
    type: 'textarea',
    placeholder: 'Раскройте тему',
    validate: { minMax: minMax(3, 2048) },
  },
] as const

export type ForumTopicFields = (typeof inputData)[number]['name']
export type ForumTopicStruct = Record<ForumTopicFields, string>

const defaultValues: ForumTopicStruct = {
  topic_name: '',
  topic_text: '',
}

export { inputData, defaultValues }
