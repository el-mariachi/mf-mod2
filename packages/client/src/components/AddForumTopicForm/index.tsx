import { useAppDispatch } from '@hooks/redux_typed_hooks'
import { addTopic } from '@store/slices/forum'
import { FC, HTMLAttributes, useState } from 'react'
import classNames from 'classnames'
import { Button, Col, Form, Row } from 'react-bootstrap'
import { SubmitHandler, useForm } from 'react-hook-form'
import FormControl from '@components/FormControl'
import SpinnerButton from '@components/SpinnerButton'
import { inputData, defaultValues, ForumTopicStruct } from './constants'
import { AppError, formUserErrorHandler } from '@utils/errorsHandling'
import { delay } from '@utils/index'
import './AddForumTopicForm.scss'

export type AddForumTopicFormProps = HTMLAttributes<HTMLDivElement> & {
  mock2topic: () => void
  mock2list: () => void
}
const AddForumTopicForm: FC<AddForumTopicFormProps> = ({
  mock2topic,
  mock2list,
  className: cls,
  ...attrs
}) => {
  const [loading, setLoading] = useState(false)
  const [readOnly, setReadOnly] = useState(false)

  const {
    register,
    handleSubmit,
    setError,
    clearErrors,
    formState: { errors },
  } = useForm<ForumTopicStruct>({
    mode: 'onTouched',
    reValidateMode: 'onChange',
    defaultValues,
  })
  const [submitError, setSubmitError] = useState('')

  const formSubmit: SubmitHandler<ForumTopicStruct> = data => {
    clearErrors()
    setLoading(true)
    setReadOnly(true)

    delay(1000)
      .then(() => {
        // TODO
      })
      .catch((error: AppError) => formUserErrorHandler(error, setSubmitError))
      .finally(() => {
        setLoading(false)
        setReadOnly(false)
        mock2topic()
      })
  }

  const [title, setTitle] = useState('')
  const dispatch = useAppDispatch()

  const handleSubmitTopic = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault()
    dispatch(addTopic(title))
  }

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    setTitle(event.target.value)
  }

  const formControls = inputData.map((controlProps, index) => (
    <FormControl
      key={index}
      formName="forumTopicForum"
      register={register}
      errors={errors}
      readOnly={readOnly}
      controlProps={controlProps}
    />
  ))

  return (
    <div className={classNames(cls, 'add-forum-topic-form')} {...attrs}>
      <h2 className="h4 fw-light mb-4">Создание новой темы</h2>
      <Form onSubmit={handleSubmit(formSubmit)}>
        {submitError ? <p className="text-danger mb-3">{submitError}</p> : null}
        {formControls}
        <Form.Group as={Row}>
          <Col sm={{ span: 9, offset: 3 }}>
            <SpinnerButton className="me-2" loading={loading}>
              Создать
            </SpinnerButton>
            <Button variant="secondary" onClick={mock2list}>
              Отменить
            </Button>
          </Col>
        </Form.Group>
      </Form>
    </div>
  )
}
export default AddForumTopicForm
