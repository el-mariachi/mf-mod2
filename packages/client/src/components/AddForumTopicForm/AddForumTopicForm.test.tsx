import { render } from '@testing-library/react'
import AddForumTopicForm from './index'

describe('#AddForumTopicForm', () => {
  it('AddForumTopicForm should to match snapshot', () => {
    const { baseElement } = render(<AddForumTopicForm />)
    expect(baseElement).toMatchSnapshot()
  })
})
