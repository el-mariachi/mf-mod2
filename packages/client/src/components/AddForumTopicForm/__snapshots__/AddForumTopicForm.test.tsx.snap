// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#AddForumTopicForm AddForumTopicForm should to match snapshot 1`] = `
<body>
  <div>
    <div
      class="overflow-hidden p-1 pe-0"
    >
      <form
        class="add-topic-form form d-flex gap-2 float-end"
      >
        <div>
          <input
            class="form-control"
            id="formNewTopic"
            placeholder="Название новой темы"
            type="text"
          />
        </div>
        <button
          class="btn btn-primary"
          type="submit"
        >
          Создать тему
        </button>
      </form>
    </div>
  </div>
</body>
`;
